### PR TITLE
Connection pooling

### PR DESCRIPTION
**Pull request by ingwarsw**
## Description
Resolved all problems with mysql connections
- Connection pooling
- Max connections
- Closing connections after web requests

## Motivation and Context
After running server some time connected to MySQL database connections was lost.
Also connections was slow (while reconnecting with each request)
Fixes:
- #1735 
- #2412 
- #1861

## How Has This Been Tested?
I tested it on server on heroku.
Tested with SqlLite and MySql.

## Screenshots (if appropriate):
- None

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kevinpark1217** [comment]
@ingwarsw Can you restrict it to a single connection? Would there be a significant performance difference between a 5 connections pooling and a single one? I believe that single connection pooling will also fix deadlock issues in MySQL.

**Comment by ingwarsw** [comment]
@kevinpark1217 I suppose not..
At least 2 connections are required.. one for search thread (which dont close connection at all) .. and second for web threads..

**Comment by kevinpark1217** [comment]
@ingwarsw I like the way it is. But, if I only run webserver on one server and search thread on the other server, then each would only require a single connection?

**Comment by ingwarsw** [comment]
Yes they should.. but didnt test it..

**Comment by jangrewe** [comment]
We had a little discussion about this a couple of days ago, and i think the consensus was that we'd prefer to have persistent connections, instead of constantly building new ones up and then tearing them down immediately, only to start over again.

http://charlesleifer.com/blog/managing-database-connections-with-peewee/

File: pogom/utils.py [hunk 1]
```diff
@@ -100,6 +100,7 @@ def get_args():
     parser.add_argument('--db-user', help='Username for the database')
     parser.add_argument('--db-pass', help='Password for the database')
     parser.add_argument('--db-host', help='IP or hostname for the database')
+    parser.add_argument('--db-max_connections', help='IP or hostname for the database', default=5)
```
Comment (invisiblek) [inline pogom/utils.py]: the help text is inaccurate here

Comment (ingwarsw) [inline pogom/utils.py]: Fixed.. along with type of variable..

**Comment by AHAAAAAAA** [comment]
Testing this out now, reporting back in a bit.

**Comment by AHAAAAAAA** [comment]
can you rebase?

**Comment by ingwarsw** [comment]
Sure.. I need to merge dropping tables also..

**Comment by ingwarsw** [comment]
Done..

**Comment by AHAAAAAAA** [comment]
Seems to be better but I still get:
`ValueError: Exceeded maximum connections.` which may just be the AWS DB I'm using limiting me.

**Comment by ingwarsw** [comment]
Whats your limit and how many search threads you have?

Im just working to not have connection per thread..

**Comment by AHAAAAAAA** [comment]
Getting limited to 24, when my instance allows 36. Anyway, I fixed it on my own DB and it seems to be working for me so +1!

**Comment by ingwarsw** [comment]
Cool. If i will find way of lovering to one for all search threads I will try pushing next PR..

**Comment by ingwarsw** [comment]
I see in pewee that every thread need its own connection.. so it will be hard to go below the search threads number.

**Comment by kevinpark1217** [comment]
My Test Log
1. One web server and one single threaded scanner. 2 total connections to the database. Passed!
2. One web server and one multithreaded scanner. 2 total connections to the database. Passed!

**Comment by kevinpark1217** [comment]
@ingwarsw @AHAAAAAAA So far it is working just as expected. Now, I am going to look for ways to use multiple workers, but use single connection to the database! Great work!

**Comment by ingwarsw** [comment]
I just added releasing connections for search threads..

Now looks even beter..

**Comment by ingwarsw** [comment]
Could you check if it works for SQLite..
I dont have more time today..

**Comment by AHAAAAAAA** [comment]
Nice, I'll test as soon as the servers are back.

**Comment by kevinpark1217** [comment]
@ingwarsw @AHAAAAAAA So it appears that each thread needs its own db connection. Am I right? Because I was able to run single thread with one db connection, but there were max connection issues with 4 threads. However, I was able to fix it by increasing the max db connections to 4.

Is it possible to improve upon here, so multiple threads can use a single db connection?
If not, then you should remove --db-max_connections, instead it should be calculated based on the number of threads user specify.
For example, -t 4 would result in 4 db connections with -ns flag, but 5 db connections with the webserver.

**Comment by FlarumHostingCo** [comment]
I'm using 2 servers, each with it's own map key and runserver.py   Will MySQL spawn pokemon and gyms on both maps based on scans from both servers if both servers are sharing one MySQL database?

**Comment by wcastello** [comment]
Testing here too.

@kevinpark1217 would using a single connection really improve anything? If I'm using lots of threads wouldn't they have to wait on the connection then?

**Comment by kevinpark1217** [comment]
@wcastello MySQL is designed to be able to handle simultaneous, so I don't think it will bottleneck. Limiting it to single connection would eliminate ports being wasted, thus, allow more database connections.

**Comment by kevinpark1217** [comment]
@ingwarsw It seems like the webserver alone is asking for more than one database connection. This should be fixed so that the webserver only requires a single db connection.

**Comment by kevinpark1217** [comment]
@ingwarsw Also, I can't figure out why does webserver alone need at least 2 MySQL connections?

**Comment by AHAAAAAAA** [comment]
@kevinpark1217 Server needs to read from database. Search thread needs to write to database. 2 connections.

Also, multiple connections can (and do) access the database from the same port. My issue, typically when gathering large amounts of data across multiple workers, is hitting max_connection limits on wherever I'm hosting my DB. I'm not exactly sure what you mean by limiting it to a single thread. Multi-threading was introduced as a feature, and it's pretty stable in terms of functionality. The usual culprits when it fails are either Niantic's servers or DB locking.

**Comment by kevinpark1217** [comment]
@AHAAAAAAA "Server needs to read from database. Search thread needs to write to database. 2 connections." Yeah I realize this. I am not some MySQL newbie, but I am only running a web service not a scanner; therefore, it should only require a single db connection, however, I am getting 2 connections.

**Comment by kevinpark1217** [comment]
@AHAAAAAAA Also, I am not against multithreading, but I believe that scanner should use a same single database connection for writing purposes. But, current pooling seems to initiate a new db connection when a thread start. Instead, the runserver should initiate database connection once, and then pass the same variable to all threads to eliminate wasted ports.

**Comment by wcastello** [comment]
Not working with sqlite:
```
Exception in thread search_thread-0:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/PokemonGo-Map/pogom/search.py", line 159, in search_thread
    parse_map(response_dict, i, step, step_location)
  File "/PokemonGo-Map/pogom/models.py", line 315, in parse_map
    bulk_upsert(Pokemon, pokemons)
  File "/PokemonGo-Map/pogom/models.py", line 353, in bulk_upsert
    InsertQuery(cls, rows=data.values()[i:min(i+step, num_rows)]).upsert().execute()
  File "/pyv2/lib/python2.7/site-packages/peewee.py", line 3244, in execute
    cursor = self._execute()
  File "/pyv2/lib/python2.7/site-packages/peewee.py", line 2656, in _execute
    return self.database.execute_sql(sql, params, self.require_commit)
  File "/pyv2/lib/python2.7/site-packages/peewee.py", line 3492, in execute_sql
    self.commit()
  File "/pyv2/lib/python2.7/site-packages/peewee.py", line 3316, in __exit__
    reraise(new_type, new_type(*exc_args), traceback)
  File "/pyv2/lib/python2.7/site-packages/peewee.py", line 3483, in execute_sql
    cursor = self.get_cursor()
  File "/pyv2/lib/python2.7/site-packages/peewee.py", line 3431, in get_cursor
    return self.get_conn().cursor()
ProgrammingError: SQLite objects created in a thread can only be used in that same thread.The object was created in thread id 140735232671744 and this is thread id 123145308196864
```

**Comment by kevinpark1217** [comment]
@ingwarsw I think it's requiring two database connections just for the webserver because inside the models.py (def init_database), it creates db and flaskDb. So, it is initializing two database connection when it could be eliminated to use one.

**Comment by kevinpark1217** [comment]
@ingwarsw Sweet I found a fix for a 2 database connections only for the webserver. Can you give me a collaborator permission so I can push it.

**Comment by AHAAAAAAA** [comment]
@kevinpark1217 make a PR, we don't directly push.

**Comment by invisiblek** [comment]
@AHAAAAAAA i think he's asking @ingwarsw for access to his repo to push on top of this branch and keep it all in the same PR

**Comment by kevinpark1217** [comment]
@invisiblek I'm just going to make a new PR that includes his since he didn't do anything for hours.

**Comment by ingwarsw** [comment]
I was sleeping :) 
I know it shouldn't happen but it does..

Pooling for Sqlite in broken and I have to fix it.. 

@kevinpark1217 what's the resolution?

**Comment by ingwarsw** [comment]
@kevinpark1217

flaskDb is just wrapper for db with adds flask app hooks (for opening/closing connections)

 Connection pooling looks like that:
- You create pool with max size and 0 ready connections.
- When you do connect on pool you take first waiting connection (or create new if there is none ready)
- When you do close on pool you release connection to waiting pool.

So when you create 10 search threads and 10 web threads there is really big chance that they will like to have more than 1 connection in same time.. 
But its also unlike that they will go over 5 (while when not pooling they would use 20).

**Comment by ingwarsw** [comment]
@wcastello fixed issue with SQLite

**Comment by kevinpark1217** [comment]
@ingwarsw I was wrong. I had some code flaws that I did not realize until a full size test.

Anyway, if I run a single instance of a webserver, how many web threads does it create? Does it depends on the number of connections to the website?

Thanks for the help!

**Comment by kevinpark1217** [comment]
@ingwarsw Also, does that mean that it will be impossible to create a global map with tons of users contributing a single database, since it can only allow up to 65535 ports?

**Comment by ingwarsw** [comment]
@kevinpark1217 Everything uses pools (so it will be ok).. and if you will have that many users you will plainly die..

**Comment by jangrewe** [comment]
@kevinpark1217 For a global map you'd probably rather use an API for ingesting all the data, rather than giving everybody write access to your database. That API would probably use a connection pool, too.

**Comment by kevinpark1217** [comment]
@invisiblek It would be great if this could be merged soon as possible.

**Comment by ingwarsw** [comment]
Heck it.. I suppose I have to do merge once more because problems with travis..

**Comment by Mr-Groch** [comment]
I have exception "ValueError: Exceeded maximum connections" on default value of 5, after first web page visit (or refresh). I have raised limit to 10 and now is OK. I'm using MySQL (MariaDB) local connection - no limits on MySQL server side.

**Pull request merged**